### PR TITLE
Ensure we always have an application name

### PR DESCRIPTION
**Pull request by SimonRichardson**
The following makes more of an effort to get an application name for the
error. When we don't have an application fallback to the charm meta
name.

The code is very simple.

## QA steps

It expects that you've run `charmcraft build` on the `juju-qa-test-v2`.

```sh
$ juju bootstrap microk8s
$ juju deploy ./testcharms/charm-hub/charms/juju-qa-test-v2/juju-qa-test-v2.charm 
```

## Bug reference

https://bugs.launchpad.net/juju/+bug/1925714

### CODE REVIEW

**Review by hmlanigan — APPROVED** [review]
QA is good

**Review by jameinel — COMMENTED** [review]
Is this something we could add a unit test for?

File: cmd/juju/application/deployer/deployer_test.go [hunk 1]
```diff
@@ -328,6 +331,47 @@ func (s *deployerSuite) TestValidateResourcesNeededForLocalDeployCAASWithIAAS(c
...
+	f := &factory{
+		clock:           clock.WallClock,
+		applicationName: "meshuggah",
+		charmOrBundle:   "local:meshuggah",
+		charmReader:     s.charmReader,
+		model:           s.modelCommand,
+	}
+
+	_, err := f.maybeReadLocalCharm(s.modelConfigGetter)
+	c.Assert(err, gc.ErrorMatches, `cannot add application "meshuggah": non container-based charm for container based model type not valid`)
```
Comment (manadart) [inline cmd/juju/application/deployer/deployer_test.go:352]: "container-based" should replace "container based" here for consistency.

**Review by manadart — APPROVED** [review]
Good change.

**Comment by SimonRichardson** [comment]
$$merge$$

**Comment by SimonRichardson** [comment]
> TestRemoveModel

$$merge$$

**Pull request merged**